docs/source/miscellaneous/stubs.rst

```diff
@@ -0,0 +1,610 @@
+For documentation purposes, PHPDoc can be used.
+
+Since PHP 8.0, arginfo structures can also store default values in order to be used by
+``ReflectionParameter::getDefaultValue()`` among some other use-cases. Default values may not only
+be literals, but compile-time evaluable expressions, possibly containing references to constants.
+Let's modify function ``foo()`` slightly in our original example by making ``$param`` an optional
+parameter:
+
+.. code:: php
+
+   function foo(string $param = FOO . "bar"): string {}
+
+This will result in the following arginfo:
+
+.. code:: c
+
+   ZEND_BEGIN_ARG_WITH_RETURN_TYPE_INFO_EX(arginfo_foo, 0, 0, IS_STRING, 0)
+      ZEND_ARG_TYPE_INFO_WITH_DEFAULT_VALUE(0, param, IS_LONG, 0, "FOO . \"bar\"")
+   ZEND_END_ARG_INFO()
+
+By default, referencing constants works as long as the constant is available in the same stub file.
+If this is not possible by any reason, then the stub declaring the necessary constant should be
+included:
+
+.. code:: php
+
+   // constants.stub.php
+
+   /** @var string */
+   const FOO = "foo";
+   /** @var string */
+   const BAR = "bar";
+
+.. code:: php
+
+   // example.stub.php
+
+   require "constants.stub.php";
+
+   function foo(string $param = FOO): string {}
+
+Sometimes, arguments have to be passed by reference, or by using the `ZEND_SEND_PREFER_REF` flag.
+Passing by reference is trivial to declare by the usual syntax, while the latter can be achieved by
+using the ``@prefer-ref`` PHPDoc tag:
+
+.. code:: php
+
+   /**
+    * @param string $param1
+    * @prefer-ref $param2
+    */
+   function foo(&$param1, string $param2): string {}
+
+*****************************
+ Generating function entries
+*****************************
+
+Besides arginfo structures, function entries themselves can also be generated via stubs. In order to
+make this work, the file-level ``@generate-function-entries`` PHPDoc tag has to be added to our
+stub:
+
+.. code:: php
+
+   /** @generate-function-entries */
+
+   class Foo {
+       public function bar(): string {}
+   }
+
+   function foo(string $param): string {}
+
+Now, the following C code is added to our original arginfo file:
+
+.. code:: c
+
+   /* ... */
+
+   ZEND_FUNCTION(foo);
+   ZEND_METHOD(Foo, bar);
+
+   static const zend_function_entry ext_functions[] = {
+       ZEND_FE(foo, arginfo_foo)
+       ZEND_FE_END
+   };
+
+   static const zend_function_entry class_Foo_methods[] = {
+       ZEND_ME(Foo, bar, arginfo_class_Foo_bar, ZEND_ACC_PUBLIC)
+       ZEND_FE_END
+   };
+
+The ``ext_functions`` variable is to be passed for the ``functions`` member of `zend_module_entry`,
+while ``class_Foo_methods`` can be used when registering the ``Foo`` class:
+
+.. code:: c
+
+   INIT_CLASS_ENTRY(ce, "Foo", class_Foo_methods);
+
+Function entries may make use of extra meta information passed via the following PHPDoc tags:
+
+-  ``@deprecated``: Triggers the usual deprecation notice when the function/method is called.
```

Comment:
Mention constants?

Reply:
Classes and constants are covered a bit below when `@generate-class-entries`  is discussed. This section is about function entries only, so I didn't want to include constants.

Reply:
ACK